src/benchmarklib/table_builder.hpp

```diff
@@ -0,0 +1,240 @@
+
+    // Reserve some space in the vectors
+    hana::for_each(_data_vectors, [&](auto&& vector) { vector.reserve(_estimated_rows_per_chunk); });
+    hana::for_each(_is_null_vectors, [&](auto&& is_null_vector) {
+      if constexpr (std::decay_t<decltype(is_null_vector)>::has_value) {
+        is_null_vector.value().reserve(_estimated_rows_per_chunk);
+      }
+    });
+  }
+
+  std::shared_ptr<Table> finish_table() {
+    if (_current_chunk_row_count() > 0) {
+      _emit_chunk();
+    }
+
+    return _table;
+  }
+
+  // Types == DataTypes, we deduce again to get universal references, so we can call with lvalues
+  template <typename... Types>
+  void append_row(Types&&... values) {
+    auto values_tuple = hana::make_tuple(std::forward<decltype(values)>(values)...);
+    BOOST_HANA_CONSTANT_ASSERT(hana::size(hana::tuple<DataTypes...>()) == hana::size(values_tuple));
+
+    // Create tuples ([&data_vector0, &is_null_vector0, value0], [&data_vector1, &is_null_vector1, value1], ...)
+    auto data_vectors_and_is_null_vectors_and_values = hana::zip_with(
+        [](auto& data_vector, auto& is_null_vector, auto&& value) {
+          return hana::make_tuple(std::reference_wrapper(data_vector), std::reference_wrapper(is_null_vector),
+                                  std::forward<decltype(value)>(value));
+        },
+        _data_vectors, _is_null_vectors, values_tuple);
+
+    // Add the values to their respective data vector
+    hana::for_each(data_vectors_and_is_null_vectors_and_values, [](auto&& data_vector_and_is_null_vector_and_value) {
+      auto& data_vector = data_vector_and_is_null_vector_and_value[hana::llong_c<0>].get();
+      auto& is_null_vector = data_vector_and_is_null_vector_and_value[hana::llong_c<1>].get();
+      auto& maybe_optional_value = data_vector_and_is_null_vector_and_value[hana::llong_c<2>];
+
+      constexpr bool column_is_nullable = std::decay_t<decltype(is_null_vector)>::has_value;
+      auto value_is_null = is_null(maybe_optional_value);
+
+      DebugAssert(column_is_nullable || !value_is_null, "cannot insert null value into not-null-column");
+
+      if (value_is_null) {
+        data_vector.emplace_back();
+      } else {
+        // on failure: make sure append_row is called with the same types that were passed to table_builder constructor
+        data_vector.emplace_back(std::move(get_value(maybe_optional_value)));
+      }
+
+      if constexpr (column_is_nullable) {
+        is_null_vector.value().emplace_back(value_is_null);
+      }
+    });
+
+    if (_current_chunk_row_count() >= _table->max_chunk_size()) {
+      _emit_chunk();
+    }
+  }
+
+ private:
+  std::shared_ptr<Table> _table;
+  UseMvcc _use_mvcc;
+  size_t _estimated_rows_per_chunk;
+
+  hana::tuple<std::vector<get_value_type<DataTypes>>...> _data_vectors;
+  hana::tuple<optional_constexpr<std::vector<bool>, (is_optional<DataTypes>())>...> _is_null_vectors;
```

Comment:
If this get's us rid of some template programming I'd be cool with leaving the optional_constexpr out here and just have an empty `std::vector<bool>` here.

Reply:
We would need an std::optional<std::vector>> to check whether a column is nullable. Basically all if constexprs would be normal ifs, which is especially bad for the one in append_row, since that's our "hot function". I will benchmark and see what the impact would be.

Reply:
ah, no need to bother, was just a thought.